### PR TITLE
Extend DSE for Env Vars along with Cmd Args

### PR DESCRIPTION
**Pull request by srivatsankrishnan**
## Summary
Lot of workloads defined in the test toml also requires sweeps in environment variables (`extra_env_vars`). Previous support for only exposed the `cmd_args` as DSE'ble. This PR expands the scope to parameters defined in `extra_env_vars` field defined in Test toml.

Similar to how users can specify the any parameters defined under `cmd_args` as list which will be activate the DSE. Likewise, they can also expand any parameters in `extra_env_vars` as list and CloudAI will activate the DSE for these parameters and add them to the agents action space.

Example

```
name = "dse_nemo_run_llama3_8b"
description = "dse_nemo_run_llama3_8b"
test_template_name = "NeMoRun"

[cmd_args]
docker_image_url = "nvcr.io/nvidia/nemo:24.12.rc3"
task = "pretrain"
recipe_name = "llama3_8b"
...
...

[extra_env_vars]
NCCL_P2P_NET_CHUNKSIZE = ["2097152", "4194304"]
NVTE_FUSED_ATTN = ["1", "0"]
```

Note:
* The global env vars is defined per System toml in CloudAI. So focussing on `extra_env_vars` which is controlled by users in Test toml in this PR. If there is a requirement for system toml, we should be able to figure something out too.
* cmd_args was a nice object. extra_env_args was defined as a dict from the get go. Lets revisit this and see if it makes sense to move towards object modulo user experience requirements. This is largely internal change and can be planned or done in a phased manner.

## Auxiliary inconsistency fixes
It looks like most of the local dev uses Python 3.10. The Github CI uses python 3.9. One corner cases is when the local CI/CD passes and the upstream ones fail.

Example is the failed CI test due to a python 3.10 feature.

In python 3.10 this is a valid way to define things instead of `Union` from typing. Though this is fixed to use Union to ensure backward compatibility, in general saves lot of time to move this CI flow to 3.10 since this is what we use in private repo and as well as most clusters.

```
Dict[str, str | List[str]]
```

```
Dict[str, Union[str, List[str]]]
```

In 3.9, this will result in the following error

```
src/cloudai/workloads/jax_toolbox/slurm_command_gen_strategy.py:82: in JaxToolboxSlurmCommandGenStrategy
    self, env_vars: Dict[str, str | List[str]], cmd_args: Dict[str, Any], num_nodes: int
E   TypeError: unsupported operand type(s) for |: 'type' and '_GenericAlias'
Error: Process completed with exit code 4.
```
@amaslenn FYI

Discussion on Match14th (with @amaslenn @TaekyungHeo ): Some older cluster might need 3.9. But cloudaix already uses 3.10. So not sure why are can't upgrade it given even cloudaix need to run on older cluster? 

But need broader discussion on this separately as agreed.

## Test Plan
* CI/CD
* Dry-run

### Nemo2.0 Llama3-8b Model
```
cloudai dry-run --system-config ../cloudaix/conf/common/system/xxxx.toml --tests-dir conf/common/test --test-scenario conf/common/test_scenario/dse_nemo_run_llama3_8b.toml 
[INFO] System Name: xxxx
[INFO] Scheduler: slurm
[INFO] Test Scenario Name: nemo_run_llama3_8b
[INFO] Checking if test templates are installed.
[INFO] Test Scenario: nemo_run_llama3_8b

Section Name: dse_nemo_run_llama3_8b_1
  Test Name: dse_nemo_run_llama3_8b
  Description: dse_nemo_run_llama3_8b
  No dependencies
[INFO] Initializing Runner [DRY-RUN] mode
[INFO] Creating SlurmRunner
[INFO] Running step 0 with action {'extra_env_vars.NCCL_P2P_NET_CHUNKSIZE': '2097152', 'extra_env_vars.NVTE_FUSED_ATTN': '1', ...obfuscated...}
[INFO] Starting test scenario execution.
[INFO] Starting test: dse_nemo_run_llama3_8b_1
[INFO] Running test: dse_nemo_run_llama3_8b_1
[INFO] Submitted slurm job: 0
[INFO] Job completed: dse_nemo_run_llama3_8b_1
[INFO] Step 1: Observation: [34.15068181818182], Reward: 0.029281992240272052
[INFO] Running step 0 with action {'extra_env_vars.NCCL_P2P_NET_CHUNKSIZE': '2097152', 'extra_env_vars.NVTE_FUSED_ATTN': '1', ...obfuscated...}
[INFO] Starting test scenario execution.
[INFO] Starting test: dse_nemo_run_llama3_8b_1
[INFO] Running test: dse_nemo_run_llama3_8b_1
[INFO] Submitted slurm job: 0
[INFO] Job completed: dse_nemo_run_llama3_8b_1
[WARNING] Skipping 'results/nemo_run_llama3_8b/dse_nemo_run_llama3_8b_1/0/2', can't handle with strategy=<class 'cloudai.workloads.nemo_run.report_generation_strategy.NeMoRunReportGenerationStrategy'>.
[INFO] Step 2: Observation: [-1.0], Reward: -1.0
[INFO] Running step 0 with action {'extra_env_vars.NCCL_P2P_NET_CHUNKSIZE': '4194304', 'extra_env_vars.NVTE_FUSED_ATTN': '1', ...obfuscated...}
[INFO] Starting test scenario execution.
[INFO] Starting test: dse_nemo_run_llama3_8b_1
[INFO] Running test: dse_nemo_run_llama3_8b_1
[INFO] Submitted slurm job: 0
[INFO] Job completed: dse_nemo_run_llama3_8b_1
[WARNING] Skipping 'results/nemo_run_llama3_8b/dse_nemo_run_llama3_8b_1/0/3', can't handle with strategy=<class 'cloudai.workloads.nemo_run.report_generation_strategy.NeMoRunReportGenerationStrategy'>.
[INFO] Step 3: Observation: [-1.0], Reward: -1.0
[INFO] Running step 0 with action {'extra_env_vars.NCCL_P2P_NET_CHUNKSIZE': '4194304', 'extra_env_vars.NVTE_FUSED_ATTN': '0', ...obfuscated...}
[INFO] Starting test scenario execution.
[INFO] Starting test: dse_nemo_run_llama3_8b_1
[INFO] Running test: dse_nemo_run_llama3_8b_1
[INFO] Submitted slurm job: 0
[INFO] Job completed: dse_nemo_run_llama3_8b_1
[WARNING] Skipping 'results/nemo_run_llama3_8b/dse_nemo_run_llama3_8b_1/0/4', can't handle with strategy=<class 'cloudai.workloads.nemo_run.report_generation_strategy.NeMoRunReportGenerationStrategy'>.
[INFO] Step 4: Observation: [-1.0], Reward: -1.0
[INFO] Running step 0 with action {'extra_env_vars.NCCL_P2P_NET_CHUNKSIZE': '2097152', 'extra_env_vars.NVTE_FUSED_ATTN': '1', ...obfuscated...}
[INFO] Starting test scenario execution.
[INFO] Starting test: dse_nemo_run_llama3_8b_1
[INFO] Running test: dse_nemo_run_llama3_8b_1
[INFO] Submitted slurm job: 0
[INFO] Job completed: dse_nemo_run_llama3_8b_1
[WARNING] Skipping 'results/nemo_run_llama3_8b/dse_nemo_run_llama3_8b_1/0/5', can't handle with strategy=<class 'cloudai.workloads.nemo_run.report_generation_strategy.NeMoRunReportGenerationStrategy'>.
[INFO] Step 5: Observation: [-1.0], Reward: -1.0
[INFO] Running step 0 with action {'extra_env_vars.NCCL_P2P_NET_CHUNKSIZE': '2097152', 'extra_env_vars.NVTE_FUSED_ATTN': '0', ...obfuscated...}
[INFO] Starting test scenario execution.
[INFO] Starting test: dse_nemo_run_llama3_8b_1
[INFO] Running test: dse_nemo_run_llama3_8b_1
[INFO] Submitted slurm job: 0
[INFO] Job completed: dse_nemo_run_llama3_8b_1
[WARNING] Skipping 'results/nemo_run_llama3_8b/dse_nemo_run_llama3_8b_1/0/6', can't handle with strategy=<class 'cloudai.workloads.nemo_run.report_generation_strategy.NeMoRunReportGenerationStrategy'>.
[INFO] Step 6: Observation: [-1.0], Reward: -1.0
[INFO] Running step 0 with action {'extra_env_vars.NCCL_P2P_NET_CHUNKSIZE': '4194304', 'extra_env_vars.NVTE_FUSED_ATTN': '1', ...obfuscated...}
[INFO] Starting test scenario execution.
[INFO] Starting test: dse_nemo_run_llama3_8b_1
[INFO] Running test: dse_nemo_run_llama3_8b_1
[INFO] Submitted slurm job: 0
[INFO] Job completed: dse_nemo_run_llama3_8b_1
[WARNING] Skipping 'results/nemo_run_llama3_8b/dse_nemo_run_llama3_8b_1/0/7', can't handle with strategy=<class 'cloudai.workloads.nemo_run.report_generation_strategy.NeMoRunReportGenerationStrategy'>.
[INFO] Step 7: Observation: [-1.0], Reward: -1.0
[INFO] Running step 0 with action {'extra_env_vars.NCCL_P2P_NET_CHUNKSIZE': '4194304', 'extra_env_vars.NVTE_FUSED_ATTN': '0', ...obfuscated...}
[INFO] Starting test scenario execution.
[INFO] Starting test: dse_nemo_run_llama3_8b_1
[INFO] Running test: dse_nemo_run_llama3_8b_1
[INFO] Submitted slurm job: 0
[INFO] Job completed: dse_nemo_run_llama3_8b_1
[WARNING] Skipping 'results/nemo_run_llama3_8b/dse_nemo_run_llama3_8b_1/0/8', can't handle with strategy=<class 'cloudai.workloads.nemo_run.report_generation_strategy.NeMoRunReportGenerationStrategy'>.
[INFO] Step 8: Observation: [-1.0], Reward: -1.0
```
### NCCL Test All-Gather

```
$ cloudai dry-run --system-config ../cloudaix/conf/common/system/xxxx.toml --tests-dir conf/common/test --test-scenario conf/common/test_scenario/dse_nccl_all_gather.toml 
[INFO] System Name: xxx
[INFO] Scheduler: slurm
[INFO] Test Scenario Name: dse-nccl-test
[INFO] Checking if test templates are installed.
[INFO] Test Scenario: dse-nccl-test

Section Name: Tests.1
  Test Name: dse_nccl_all_gather
  Description: all_gather
  No dependencies
[INFO] Initializing Runner [DRY-RUN] mode
[INFO] Creating SlurmRunner
[INFO] Running step 0 with action {'docker_image_url': 'nvcr.io/nvidia/pytorch:24.02-py3', 'subtest_name': 'all_gather_perf_mpi', 'nthreads': 1, 'ngpus': 1, 'minbytes': '128', 'maxbytes': '4G', 'stepbytes': '1M', 'op': 'sum', 'datatype': 'float', 'root': 0, 'iters': 100, 'warmup_iters': 50, 'agg_iters': 1, 'average': 1, 'parallel_init': 0, 'check': 1, 'blocking': 0, 'cudagraph': 0, 'extra_env_vars.NCCL_TESTS_SPLIT_MASK': '0x7'}
[INFO] Starting test scenario execution.
[INFO] Starting test: Tests.1
[INFO] Running test: Tests.1
[INFO] Submitted slurm job: 0
[INFO] Job completed: Tests.1
[WARNING] Skipping 'results/dse-nccl-test/Tests.1/0/1', can't handle with strategy=<class 'cloudai.workloads.nccl_test.report_generation_strategy.NcclTestReportGenerationStrategy'>.
[INFO] Step 1: Observation: [-1.0], Reward: -1.0
[INFO] Running step 0 with action {'docker_image_url': 'nvcr.io/nvidia/pytorch:24.02-py3', 'subtest_name': 'all_gather_perf_mpi', 'nthreads': 1, 'ngpus': 1, 'minbytes': '128', 'maxbytes': '4G', 'stepbytes': '1M', 'op': 'sum', 'datatype': 'float', 'root': 0, 'iters': 100, 'warmup_iters': 50, 'agg_iters': 1, 'average': 1, 'parallel_init': 0, 'check': 1, 'blocking': 0, 'cudagraph': 0, 'extra_env_vars.NCCL_TESTS_SPLIT_MASK': '0x0'}
[INFO] Starting test scenario execution.
[INFO] Starting test: Tests.1
[INFO] Running test: Tests.1
[INFO] Submitted slurm job: 0
[INFO] Job completed: Tests.1
[WARNING] Skipping 'results/dse-nccl-test/Tests.1/0/2', can't handle with strategy=<class 'cloudai.workloads.nccl_test.report_generation_strategy.NcclTestReportGenerationStrategy'>.
[INFO] Step 2: Observation: [-1.0], Reward: -1.0
```

* Real system testing
### NCCL AG with Extra Env Args
```
cloudai run --system-config ../cloudaix/conf/common/system/c
xxx --tests-dir conf/common/test --test-scenario conf/common/test_scenario/dse_nccl_all_gather.toml 
[INFO] System Name: xxx
[INFO] Scheduler: slurm
[INFO] Test Scenario Name: dse-nccl-test
[INFO] Checking if test templates are installed.
[INFO] Test Scenario: dse-nccl-test

Section Name: Tests.1
  Test Name: dse_nccl_all_gather
  Description: all_gather
  No dependencies
[INFO] Initializing Runner [RUN] mode
[INFO] Creating SlurmRunner
[INFO] Running step 0 with action {...obfuscated...,'extra_env_vars.NCCL_TESTS_SPLIT_MASK': '0x7'}
[INFO] Starting test scenario execution.
[INFO] Starting test: Tests.1
[INFO] Running test: Tests.1
[INFO] Submitted slurm job: 1881080
[INFO] Job completed: Tests.1
[INFO] Step 1: Observation: [-1.0], Reward: -1.0
[INFO] Running step 0 with action {...obfuscated.... 'extra_env_vars.NCCL_TESTS_SPLIT_MASK': '0x0'}
[INFO] Starting test scenario execution.
[INFO] Starting test: Tests.1
[INFO] Running test: Tests.1
[INFO] Submitted slurm job: 1881207
[INFO] Job completed: Tests.1
[INFO] Step 2: Observation: [-1.0], Reward: -1.0
```
Output (representative). Please ping me offline for more details.

image

### NCCL AG with Extra Env Args and CmdArgs

```
cloudai run --system-config ../cloudaix/conf/common/system/xxxx.to
ml --tests-dir conf/common/test --test-scenario conf/common/test_scenario/dse_nccl_all_gather.toml 
[INFO] System Name: xxxx
[INFO] Scheduler: slurm
[INFO] Test Scenario Name: dse-nccl-test
[INFO] Checking if test templates are installed.
[INFO] Test Scenario: dse-nccl-test

Section Name: Tests.1
  Test Name: dse_nccl_all_gather
  Description: all_gather
  No dependencies
[INFO] Initializing Runner [RUN] mode
[INFO] Creating SlurmRunner
[INFO] Running step 0 with action {output redacted.... 'warmup_iters': 5, 'extra_env_vars.NCCL_TESTS_SPLIT_MASK': '0x7'}
[INFO] Starting test: Tests.1
[INFO] Running test: Tests.1
[INFO] Submitted slurm job: 1971248
[INFO] Job completed: Tests.1
[INFO] Generated scenario report at results/dse-nccl-test/dse-nccl-test.html
[INFO] Step 1: Observation: [-1.0], Reward: -1.0
[INFO] Running step 0 with action {output redacted...'warmup_iters', 'extra_env_vars.NCCL_TESTS_SPLIT_MASK': '0x0'}
[INFO] Starting test: Tests.1
[INFO] Running test: Tests.1
[INFO] Submitted slurm job: 1971250
[INFO] Job completed: Tests.1
[INFO] Generated scenario report at results/dse-nccl-test/dse-nccl-test.html
[INFO] Step 2: Observation: [-1.0], Reward: -1.0
[INFO] Running step 0 with action {output redacted....'warmup_iters': 50, 'extra_env_vars.NCCL_TESTS_SPLIT_MASK': '0x7'}
[INFO] Starting test: Tests.1
[INFO] Running test: Tests.1
[INFO] Submitted slurm job: 1971252
[INFO] Job completed: Tests.1
[INFO] Generated scenario report at results/dse-nccl-test/dse-nccl-test.html
[INFO] Step 3: Observation: [-1.0], Reward: -1.0
[INFO] Running step 0 with action {output redacted...'warmup_iters': 50, 'extra_env_vars.NCCL_TESTS_SPLIT_MASK': '0x0'}
[INFO] Starting test: Tests.1
[INFO] Running test: Tests.1
[INFO] Submitted slurm job: 1971317
[INFO] Job completed: Tests.1
[INFO] Generated scenario report at results/dse-nccl-test/dse-nccl-test.html
[INFO] Step 4: Observation: [-1.0], Reward: -1.0
```
Results
Representative output. Ping me on slack for more details

image

## Additional Notes
More unit test and workloads (e.g., NCCL, Nemo) will be tested. UCC will be tested after sync-up with Sergey. But can be outside of this PR. Note, the reward generation etc depends upon the objects. So the generic report will be useful. But can fix this once we understand the objective.

### CODE REVIEW

File: src/cloudai/_core/test.py [hunk 1]
```diff
@@ -134,7 +134,7 @@ class TestDefinition(BaseModel, ABC):
     description: str
     test_template_name: str
     cmd_args: Any
-    extra_env_vars: dict[str, str] = {}
+    extra_env_vars: Any = {}
```
Comment (srivatsankrishnan) [inline src/cloudai/_core/test.py:137]: @amaslenn Not sure if this is pydantic way. This a required field and without default `{}` was getting pydantic errors. Lets revist the scope keeping future directions in mind.

File: src/cloudai/_core/configurator/cloudai_gym.py [hunk 2]
```diff
@@ -215,12 +223,21 @@ def parse_report(self, output_path: Path) -> list:
     def update_nested_attr(self, obj, attr_path, value):
         """Update a nested attribute of an object."""
         attrs = attr_path.split(".")
+        # extra_env_vars is currently a dict (will remove this comment!)
```
Comment (TaekyungHeo) [inline src/cloudai/_core/configurator/cloudai_gym.py:226]: Please remove this comment before merging.

File: src/cloudai/_core/configurator/cloudai_gym.py [hunk 3]
```diff
@@ -215,12 +223,21 @@ def parse_report(self, output_path: Path) -> list:
     def update_nested_attr(self, obj, attr_path, value):
```
Comment (TaekyungHeo) [inline src/cloudai/_core/configurator/cloudai_gym.py:223]: Several comments:
1. Let's add type annotations.
2. Do we have unit tests for `update_nested_attr`? If so, please update them. If not, let's add one.
3. A minor concern - `update_nested_attr` is not very readable or maintainable. It’s okay for now if we want to stick with the current design, but we should aim for gradual improvements.

**Review by TaekyungHeo — CHANGES_REQUESTED** [review]
Let's have two approvals. Please review my comments. Some are minor.

File: src/cloudai/_core/test.py (hunk 1 above)
Comment (amaslenn) [reply src/cloudai/_core/test.py:137]: I don't get you question here. A field with default value is not required, one can create an instance of `TestDefinition` without defining `extra_env_vars`.

BTW, it shouldn't have type `Any`, I believe it should be `dict[str, Union[str, list[str]]`.

File: src/cloudai/_core/configurator/cloudai_gym.py [hunk 4]
```diff
@@ -212,15 +225,23 @@ def parse_report(self, output_path: Path) -> list:
                     return [average_value]
         return [-1.0]
 
-    def update_nested_attr(self, obj, attr_path, value):
+    def update_test_run_obj(self, obj, attr_path, value):
```
Comment (TaekyungHeo) [inline src/cloudai/_core/configurator/cloudai_gym.py:228]: 1. Please add type annotations
2. Please add unit tests for `update_test_run_obj`

**Review by TaekyungHeo — CHANGES_REQUESTED** [review]

**Review by TaekyungHeo — DISMISSED** [review]
LGTM. Let's have two approvals.

File: src/cloudai/_core/test.py (hunk 1 above)
Comment (amaslenn) [reply src/cloudai/_core/test.py:137]: We shouldn't touch `extra_cmd_args`, right? We will deprecate it soon. New type should be applied to the `extra_ENV_vars`.

File: src/cloudai/_core/configurator/cloudai_gym.py [hunk 5]
```diff
@@ -59,7 +59,15 @@ def define_action_space(self) -> Dict[str, Any]:
         """
         action_space: Dict[str, Any] = {}
         cmd_args_dict = self.test_run.test.test_definition.cmd_args.model_dump()
-        self.populate_action_space("", cmd_args_dict, action_space)
+        extra_env_vars_dict = self.test_run.test.test_definition.extra_env_vars
+
+        combined_dict = {
+            **{f"{key}": value for key, value in cmd_args_dict.items()},
```
Comment (amaslenn) [inline src/cloudai/_core/configurator/cloudai_gym.py:65]: [minor] can it be simplified to `**cmd_args_dict`?

File: src/cloudai/_core/configurator/cloudai_gym.py [hunk 6]
```diff
@@ -59,7 +59,15 @@ def define_action_space(self) -> Dict[str, Any]:
         """
         action_space: Dict[str, Any] = {}
         cmd_args_dict = self.test_run.test.test_definition.cmd_args.model_dump()
-        self.populate_action_space("", cmd_args_dict, action_space)
+        extra_env_vars_dict = self.test_run.test.test_definition.extra_env_vars
+
+        combined_dict = {
```
Comment (amaslenn) [inline src/cloudai/_core/configurator/cloudai_gym.py:64]: Do you think it make sense to have something like that:
```py
@dataclass
class ActionSpace:
  cmd_args: ...
  env_vars: ...
```

The usage will become much easier, we won't need checking the subsctr.

File: src/cloudai/_core/test.py (hunk 1 above)
Comment (srivatsankrishnan) [reply src/cloudai/_core/test.py:137]: Yeah agree. We are not touching `extra_cmd_args` in this PR and will generally not support it as well going forward.

File: src/cloudai/_core/configurator/cloudai_gym.py (hunk 6 above)
Comment (srivatsankrishnan) [reply src/cloudai/_core/configurator/cloudai_gym.py:64]: Generally makes sense. But lets unblock two users might be able to try this and provide more feedback on this. I would like more Nemo use cases to soldify our Nemo2.0 implementation first and this PR unblocks one of those use case. 

Need to think through a bit on downstream integration with more smarter agents.
* Gym interfaces makes RL first class citizens. We are just repurposing it for other agents we have so far. Depending upon frameworks/future agents for RL/learning based methods, we might need additional logic to construct/reconstruct object to dictionaries anyway. I feel leaving at dictionaries makes most sense at this point since we know it should work with most of the `spaces` types used in RL/learning based ones. 
* Lets also think about our strategy on what all we can really sweep. We can currently see `cmd_args` and `env_vars`. But if you think about it, any field in the test toml can be DSE'ble (e.g., think about how users can provide multiple docker urls and also run the same configs across two images). In that case, I see a very strong motivation to move this to class. At that point moving to having a shim layer to convert obj to dictionaries can be amortized.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -17,7 +17,7 @@ jobs:
       - name: Set up Python
         uses: actions/setup-python@v5
         with:
-          python-version: 3.9
```
Comment (amaslenn) [inline .github/workflows/ci.yml:20]: Why do we need this version bump?

File: src/cloudai/_core/configurator/grid_search.py [hunk 8]
```diff
@@ -63,7 +63,7 @@ def get_all_combinations(self) -> List[Dict[str, Any]]:
             List[Dict[str, Any]]: A list of dictionaries, each representing a unique combination of parameters.
         """
         keys = list(self.action_space.keys())
-        return [dict(zip(keys, combination)) for combination in self.action_combinations]
+        return [dict(zip(keys, combination, strict=False)) for combination in self.action_combinations]
```
Comment (amaslenn) [inline src/cloudai/_core/configurator/grid_search.py:66]: Why `strict` is needed? `len(keys) != len(combination)` might indicate and internal bug, `strict=False` might cover it and result in invalid generation.

File: src/cloudai/workloads/jax_toolbox/slurm_command_gen_strategy.py [hunk 9]
```diff
@@ -78,14 +78,17 @@ def _update_env_vars(self, tr: TestRun):
 
         tr.test.test_definition.extra_env_vars.update(env_vars)
 
-    def _update_per_gpu_combine_threshold(self, env_vars: Dict[str, str], cmd_args: Dict[str, Any], num_nodes: int):
-        combine_threshold_bytes = int(env_vars["COMBINE_THRESHOLD"])
+    def _update_per_gpu_combine_threshold(
+        self, env_vars: Dict[str, Union[str, List[str]]], cmd_args: Dict[str, Any], num_nodes: int
+    ):
+        combine_threshold = env_vars["COMBINE_THRESHOLD"]
```
Comment (amaslenn) [inline src/cloudai/workloads/jax_toolbox/slurm_command_gen_strategy.py:84]: Shouldn't it be `combine_threshold_bytes = int(env_vars["COMBINE_THRESHOLD"][0]) if isinstance(env_vars["COMBINE_THRESHOLD"], list) else int(env_vars["COMBINE_THRESHOLD"])`?

P.S. I'm afraid we might face more of such issues: in cmd gen code we _know_ that all values are scalars, but we can't tell that to type system and we such errors from `pyright`.

File: src/cloudai/workloads/nccl_test/slurm_command_gen_strategy.py [hunk 10]
```diff
@@ -29,12 +29,16 @@ def _container_mounts(self, tr: TestRun) -> list[str]:
         mounts: list[str] = []
         env = tr.test.extra_env_vars | self.system.global_env_vars
         if "NCCL_TOPO_FILE" in env:
-            nccl_topo_file = Path(env["NCCL_TOPO_FILE"]).resolve()
+            nccl_topo_file = Path(env["NCCL_TOPO_FILE"]).resolve()  # pyright: ignore [reportArgumentType]
```
Comment (amaslenn) [inline src/cloudai/workloads/nccl_test/slurm_command_gen_strategy.py:32]: Same as `env_vars["COMBINE_THRESHOLD"]`, let's try to avoid static checks disabling

File: src/cloudai/workloads/jax_toolbox/slurm_command_gen_strategy.py (hunk 9 above)
Comment (srivatsankrishnan) [reply src/cloudai/workloads/jax_toolbox/slurm_command_gen_strategy.py:84]: Discussion: Lets keep ignore for now. For future, we need to revisit this on how to avoid this for other use cases.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (amaslenn) [reply .github/workflows/ci.yml:20]: Please resolve once reverted.

File: src/cloudai/_core/configurator/grid_search.py (hunk 8 above)
Comment (amaslenn) [reply src/cloudai/_core/configurator/grid_search.py:66]: Please resolve once reverted.

Comment (srivatsankrishnan) [reply src/cloudai/_core/configurator/grid_search.py]: Confirmed: ruff version 3.9 doesn't require this `strict`. 3.10, it enforces it to have `strict=false` or some value I think. Basically ruff --fix will add this btw. FYI

**Review by TaekyungHeo — APPROVED** [review]
LGTM. Let's remove `pyright ignore` gradually.

File: src/cloudai/_core/configurator/grid_search.py [hunk 11]
```diff
@@ -73,7 +73,7 @@ def select_action(self) -> Tuple[int, Dict[str, Any]]:
             Tuple[int, Dict[str, Any]]: The current step and a dictionary mapping action keys to selected
             values.
         """
-        action = dict(zip(self.action_space.keys(), self.action_combinations[self.index]))
+        action = dict(zip(self.action_space.keys(), self.action_combinations[self.index], strict=False))
```
Comment (amaslenn) [inline src/cloudai/_core/configurator/grid_search.py:76]: Is `strict` needed here?

File: src/cloudai/workloads/nccl_test/slurm_command_gen_strategy.py (hunk 10 above)
Comment (TaekyungHeo) [inline src/cloudai/workloads/nccl_test/slurm_command_gen_strategy.py:32]: I recall that we agreed not to introduce # pyright: ignore anymore. Actually, I plan to see if I can remove any lint ignore directives. Could you please check if this is essential?

**Review by TaekyungHeo — CHANGES_REQUESTED** [review]
- Please review the comment about `# pyright: ignore`. This is the only blocking issue.
- Another minor comment, which is not a blocker: You replaced `Dict[str, str]` with `Union[str, List[str]]`. You might consider introducing a new type, such as `DSEableParam = Union[str, List[str]]`, or creating a new class to improve readability and maintainability.

File: src/cloudai/workloads/nccl_test/slurm_command_gen_strategy.py (hunk 10 above)
Comment (TaekyungHeo) [reply src/cloudai/workloads/nccl_test/slurm_command_gen_strategy.py:32]: If that is the case, we can leave it as is. However, all ignores should be removed by the end of this year at the latest. I would prefer to avoid using 'pyright: ignore.'

**Review by TaekyungHeo — APPROVED** [review]

File: src/cloudai/workloads/nccl_test/slurm_command_gen_strategy.py (hunk 10 above)
Comment (srivatsankrishnan) [reply src/cloudai/workloads/nccl_test/slurm_command_gen_strategy.py:32]: @TaekyungHeo  please feel free to put a PR to remove all the ignores in CloudAI repo @TaekyungHeo . I don;t see why it needs to wait till end of this year,

**Comment by srivatsankrishnan** [comment]
> * Minor comment, which is not a blocker: You replaced `Dict[str, str]` with `Union[str, List[str]]`. You might consider introducing a new type, such as `DSEableParam = Union[str, List[str]]`, or creating a new class to improve readability and maintainability.

I don;t think this is required for such a small change. This is basic python and no need to have a class definition. I infact think this makes it worse.

**Review by amaslenn — APPROVED** [review]

**Review by TaekyungHeo — APPROVED** [review]
